pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -669,6 +669,11 @@ func (nc *Controller) monitorNodeHealth() error {
 		glog.V(1).Infof("Controller observed a Node deletion: %v", deleted[i].Name)
 		nodeutil.RecordNodeEvent(nc.recorder, deleted[i].Name, string(deleted[i].UID), v1.EventTypeNormal, "RemovingNode", fmt.Sprintf("Removing Node %v from Controller", deleted[i].Name))
 		delete(nc.knownNodeSet, deleted[i].Name)
+		if utilfeature.DefaultFeatureGate.Enabled(features.NodeLease) {
+			if err := nc.kubeClient.CoordinationV1beta1().Leases(v1.NamespaceNodeLease).Delete(deleted[i].Name, metav1.NewDeleteOptions(0)); err != nil {
```

Comment:
This has a problem that if this fail (or if controller-manager will be killed/stopped in the meantime or sth like that), the lease object will be left.

Wouldn't it be possible to set ownerReference of the Lease object to be the corresponding Node object and rely on GC? @janetkuo - thought?

Reply:
> This has a problem that if this fail (or if controller-manager will be killed/stopped in the meantime or sth like that), the lease object will be left.

The alternative is to compare the existing nodes and leases and do the cleanup every time when `monitorNodeHealth()` runs.

> Wouldn't it be possible to set ownerReference of the Lease object to be the corresponding Node object and rely on GC? Janet Kuo - thought?

IIUC, OwnerReference approach is not feasible because an owning object must be in the same namespace as the owned object, and node does not have namespace. See [here](https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L289). I also included this in the PR description.

Reply:
> The alternative is to compare the existing nodes and leases and do the cleanup every time when monitorNodeHealth() runs.

@wangzhen127 do the leases have information about which node acquired it ?

>IIUC, OwnerReference approach is not feasible because an owning object must be in the same namespace as the owned object, and node does not have namespace. See here. I also included this in the PR description.

@janetkuo  @caesarxuchao  Does the GC work on non namespaced objects ?

Reply:
I really think that we need something that is more robust that this.
Listing all leases (from local cache should be good enough) and comparing them with existing nodes sounds reasonable.

[If we won't be able to use GC - I pinged Janet and Chao about that offline].

Reply:
GC support owner being non namespaced, while the dependents are namespaced. (see https://github.com/kubernetes/kubernetes/issues/65200).

Reply:
@wangzhen127 - if that mechanism works, then we should use it.

Reply:
SG. Will try owner reference and update later.